arrow-site/docs/docs/fx/purity-and-referentially-transparent-functions/README.MD

```diff
@@ -0,0 +1,129 @@
+---
+layout: docs-fx
+title: "Kotlin Standard Library and Arrow Fx Coroutines"
+permalink: /fx/purity-and-referentially-transparent-functions/
+---
+
+## Pure & Referentially Transparent Functions
+
+A pure function is a function that consistently returns the same output when given the same input.
+Pure functions exhibit a deterministic behavior and cause no observable effects externally. We call this property referential transparency.
+
+Referential transparency allows us to reason about the different pieces of our program in isolation.
+
+To create a pure function in Kotlin, let's use the keyword `fun`:
+
+```kotlin:ank:playground
+//sampleStart
+fun helloWorld(): String =
+  "Hello World"
+
+//sampleEnd  
+fun main() {
+  println(helloWorld())
+}
+```
+
+We can state that `helloWorld` is a pure and referentially transparent function because invoking `helloWorld()` consistently returns the same output given the same input, and does not produce observable changes in the external world.
+
+## Side effects
+
+A side effect is an externally observable effect a function performs in addition to returning a value.
+
+Performing network or file I/O, writing to streams, and, in general, all functions that return `Unit` are very likely to produce side effects. That's because a `Unit` return value denotes *no useful return*, which implies that the function does nothing but perform effects.
+
+In Arrow Fx, we use `suspend fun` to denote a function that may cause side effects when invoked.
+
+In the example below, `println(a : Any): Unit` is a side effect because, every time it's invoked, it causes observable effects by interacting with the `System.out` stream and signals that it produces no useful output by returning `Unit`.
+
+When we denote side effects as `suspend`, the Kotlin compiler will ensure that we're not applying uncontrolled side effects in the pure environment.
+
+```kotlin
+//sampleStart
+fun helloWorld(): String =
+  "Hello World"
+
+suspend fun sayHello(): Unit =
+  println(helloWorld())
+
+fun main() {
+  sayHello()
+}
+//sampleEnd  
+```
+
+Compiling the snippet above will result in `javax.script.ScriptException: error: suspend function 'sayHello' should be called only from a coroutine or another suspend function` compilation error for `sayHello` call.
+
+The Kotlin compiler disallows the invocation of suspended functions in the pure environment because `suspend fun` requires declaration inside another suspended function or a continuation.
+
+A continuation is a Kotlin Interface, `Continuation<A>`, that proves we know how to handle success and error cases resulting from running the suspended effect.
+
+This is a great built-in feature of the Kotlin compiler that already makes it an ideal choice for Typed FP, but it's not the only one.
+
+Continue reading on if you're curious to see how the Kotlin Compiler and Arrow Fx can eliminate many of the functional idioms by using direct syntax, overall effectful monads.
```

Comment:
```suggestion
Continue reading on if you're curious to see how the Kotlin Compiler and Arrow Fx can eliminate many of the functional idioms by using direct syntax, over effect monads.
```
This applies unless you have do notation or similar available and a compiler that can unbox and remove the effect monad later on :wink:

Reply:
😁